cpp/src/groupby/hash/groupby.cu

```diff
@@ -509,22 +477,22 @@ void compute_single_pass_aggs(table_view const& keys,
   auto d_sparse_table = mutable_table_device_view::create(sparse_table, stream);
   auto d_values       = table_device_view::create(flattened_values, stream);
   auto const d_aggs   = cudf::detail::make_device_uvector_async(agg_kinds, stream);
-
-  bool skip_key_rows_with_nulls = keys_have_nulls and include_null_keys == null_policy::EXCLUDE;
+  auto const skip_key_rows_with_nulls =
+    keys_have_nulls and include_null_keys == null_policy::EXCLUDE;
 
   auto row_bitmask =
```

Comment:
Vindication! This is broken because bitmask_and is not going to consider the bitmask for the struct's children.

Changing this to use flattened keys again saw the performance improve back to pre-PR levels. I still think there's one more bug because the new map size still off by one. I'll continue investigating tomorrow.